bin/generate_peptides.py

```diff
@@ -45,51 +45,92 @@ def parse_args(args=None):
     parser.add_argument('-l', '--proteins_lengths', required=True, metavar='FILE', type=argparse.FileType('w'), help='Output file containing: protein_id, protein_length.')
     return parser.parse_args(args)
 
-def gen_peptides(prot_seq, k):
-    return [ prot_seq[i:(i+k)] for i in range(len(prot_seq)-k+1) ]
+# Validate letters of input protein sequences to avoid unnoticed loss of input k-mers
+def validate_letters(string, alphabet):
+    for letter in string:
+        if letter not in alphabet:
+            print("ERROR: invalid input letter ", letter, ". The supported alphabet is ", alphabet, ".")
+            sys.exit(1)
+
+def gen_peptides(prot_seq, k, prefix):
+    return [ prot_seq[i:(i+k)] for i in range(len(prot_seq)-k+1) if prot_seq[i] == prefix ]
 
 
 def main(args=None):
     args = parse_args(args)
+    print_mem = None    # 'deep' (extra computational costs) or None
+
+    # Generate peptides in chunks based on initial AA to reduce memory usage
+    # valid amino acid codes:
+    # 20 standard ('A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y')
+    # and extended codes ('B', 'J', 'O', 'U', 'X', 'Z')
+    aa_list = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', 'B', 'J', 'O', 'U', 'X', 'Z']
 
     protid_protseq_protlen = pd.read_csv(args.proteins, sep="\t")
-    protid_protseq_protlen["protein_length"] = protid_protseq_protlen["protein_sequence"].apply(len)
+    protid_protseq_protlen["protein_sequence"] = protid_protseq_protlen["protein_sequence"].str.upper()
+    protid_protseq_protlen["protein_sequence"].apply(validate_letters, alphabet=aa_list)
```

Comment:
Just out of curiosity, does this have any noticeable runtime effect?

Reply:
Noticeable yes, but not crucial I think. Don't remember the details to be honest